homeassistant/components/scheduler/__init__.py

```diff
@@ -0,0 +1,506 @@
+            self._duration = None
+
+    def __str__(self) -> str:
+        """Define the string representation of this schedule."""
+        return (
+            f'<Schedule start="{self.start_datetime}" '
+            f'end="{self.end_datetime}" rrule="{self.rrule}">'
+        )
+
+    @property
+    def active(self) -> bool:
+        """Return whether the schedule is currently active."""
+        # Schedules with no end datetimes are one-shots and are not "active":
+        if not self.end_datetime:
+            return False
+
+        now = datetime.now()
+
+        # If we're between the start and the end time, we're active:
+        if self.start_datetime <= now <= self.end_datetime:
+            return True
+
+        # If we have a recurrence, look at the instance prior to right now and determine
+        # if we're still in it:
+        last_recurrence_start_datetime = self.rrule.before(now, inc=True)
+        last_recurrence_end_datetime = last_recurrence_start_datetime + self._duration
+        return last_recurrence_start_datetime <= now <= last_recurrence_end_datetime
+
+    def as_dict(self) -> dict:
+        """Return the schedule as a dict."""
+        the_dict = {
+            CONF_ENTITY_ID: self.entity_id,
+            CONF_START_DATETIME: self.start_datetime.isoformat(),
+        }
+
+        if self.end_datetime:
+            the_dict[CONF_END_DATETIME] = self.end_datetime.isoformat()
+        else:
+            the_dict[CONF_END_DATETIME] = None
+
+        if self.rrule:
+            the_dict[CONF_RRULE] = str(self.rrule)
+        else:
+            the_dict[CONF_RRULE] = None
+
+        return the_dict
+
+    async def async_schedule_next_instance(self) -> None:
+        """Schedule the next instance of a schedule."""
+        now = datetime.now()
+
+        # Calculate the start and end datetimes for the next instance:
+        if not self._initial_run_complete:
+            self._initial_run_complete = True
+            start_datetime = self.start_datetime
+            end_datetime = self.end_datetime
+        elif self.rrule:
+            start_datetime = self.rrule.after(now, inc=True)
+            if self.end_datetime:
+                end_datetime = start_datetime + self._duration
+            else:
+                end_datetime = None
+
+        if not start_datetime:
+            _LOGGER.debug("No more instances of schedule: %s", self)
+            return
+
+        self.active_instance: ActiveScheduleInstance = ActiveScheduleInstance(
+            self._hass,
+            self.async_schedule_next_instance,
+            self.entity_id,
+            start_datetime,
+            end_datetime=end_datetime,
+        )
+        self.active_instance.async_init()
+
+        _LOGGER.info(
+            "Scheduled instance of schedule %s (start: %s / end: %s)",
+            self,
+            start_datetime,
+            end_datetime,
+        )
+
+
+@bind_hass
+class Scheduler:
+    """A class to manage scene schedules."""
```

Comment:
I feel like a lot of design is driven by the websocket API. We shouldn't.

Instead, think of it as standalone objects. Forget about the Websocket API.

What are the responsibilities of each object:

- `Scheduler`: responsible for managing schedules: CRUD and call `Schedule.schedule()` . Anything related to adding/removing schedules: happens inside scheduler object.
- `Schedule`: holds the data of a schedule, responsible for firing at the right times and for cleaning up after itself. If you need to do calculations for a schedule (expired, next instance, is an instance active), it happens inside the schedule object.
- `ScheduleInstance`: we're live! We activated, and if there is an endtime we're observering states so we can revert stuff. Anything related to actually talking to Home Assistant: happens in `ActiveSchedule`.

Reply:
break it up in multiple files. One for the websocket API, one for schedule objects. Add tests for the individual objects, it is so easy to test your assumptions and that your code works.